lib/pipelineTemplateVersionFactory.js

```diff
@@ -203,38 +204,48 @@ class PipelineTemplateVersionFactory extends BaseFactory {
-     * @param  {Object}     config               Config object
-     * @param  {Datastore}  config.datastore     Object that will perform operations on the datastore
-     * @param  {String}     [config.templateId]  The template id
-     * @param  {String}     config.name          The template name
-     * @param  {String}     config.namespace     The template namespace
-     * @param  {String}     config.version       The template version
-     * @param  {Factory}   templateMetaFactory  Template meta factory
-     * @return {Promise}                         {PipelineTemplateVersionModel, PipelineTemplateMetaModel}
+     * @param  {Object}     config                Config object
+     * @param  {String}     config.name           The template name
+     * @param  {String}     config.namespace      The template namespace
+     * @param  {String}     [config.version]      The template version
+     * @param  {String}     [config.versionOrTag] The template version or tag
+     * @param  {Factory}    templateMetaFactory   Template meta factory
+     * @param  {Factory}    [templateTagFactory]  Template tag factory
+     * @return {Promise}                          {PipelineTemplateVersionModel, PipelineTemplateMetaModel}
      */
-    async getWithMetadata(config, templateMetaFactory) {
+    async getWithMetadata(config, templateMetaFactory, templateTagFactory) {
         let pipelineTemplateMeta = {};
+        let templateVersion = config.version;
 
-        if (config.templateId) {
-            pipelineTemplateMeta = await templateMetaFactory.get({
-                id: config.templateId
-            });
-        } else {
-            pipelineTemplateMeta = await templateMetaFactory.get({
+        pipelineTemplateMeta = await templateMetaFactory.get({
+            name: config.name,
+            namespace: config.namespace
+        });
+
+        if (!pipelineTemplateMeta) {
+            return null;
+        }
+
+        if (config.versionOrTag && EXACT_VERSION_REGEX.test(config.versionOrTag)) {
+            templateVersion = config.versionOrTag;
+        } else if (config.versionOrTag && TEMPLATE_TAG_NAME_REGEX.test(config.versionOrTag)) {
+            const templateTag = await templateTagFactory.get({
                 name: config.name,
-                namespace: config.namespace
+                namespace: config.namespace,
+                tag: config.versionOrTag
             });
 
-            if (pipelineTemplateMeta) {
-                config.templateId = pipelineTemplateMeta.id;
+            if (!templateTag) {
+                return null;
             }
-        }
 
-        if (!pipelineTemplateMeta) {
-            return null;
+            templateVersion = templateTag.version;
         }
 
-        const pipelineTemplateVersion = await super.get(config);
+        const pipelineTemplateVersion = await super.get({
+            templateId: pipelineTemplateMeta.id,
+            version: templateVersion
+        });
```

Comment:
We should not cherry pick the params

```suggestion
        const pipelineTemplateVersion = await super.get(versionGetConfig);
```

Reply:
pipelineTemplateMeta.id is the source of truth for template id no? can you elaborate a bit more?